### PR TITLE
chore: make nightly build pipeline faster

### LINKED ISSUE #1943: improve build time for Rust compilation

### PR DESCRIPTION
**Pull request by BulkBeing**
closes #1943 
Same as #1949 . Trying smaller changes at a time, since I'm not able to figure out why tests for monovertex is failing in CI #1949 
UPDATE: The CI was failing becuase the Rust binary within the image didn't had executable permission.

## Summary
- GH Actions uses Ubuntu VM with `x86_64` architecture. This results in very slow build for `arm64` binary within buildkit's emulation.
- This PR moves Rust builds out of buildkit. The binaries will be built natively in the GH Actions VM in separate jobs.
- These binaries will be made available within the `dist/` directory when `numaflow` container image is built (same as the pattern we follow for Go binary).
- Uses [cargo-chef](https://github.com/LukeMathWalker/cargo-chef) to build the image on laptop. This avoids recreating Rust project directory structure in `Dockerfile` if we add/remove a cargo workspace.
- [sccache](https://github.com/mozilla/sccache/) is not used in the `release` pipeline.
- The `image` target in `Makefile` is modified to work in both `amd64` and `arm64` hosts.

These changes should reduce the nightly builds time from current average of 1hr 30 minutes to around 5 minutes.
In my testing of the nightly build pipeline, a fresh build completes in under 7 minutes. Subsequent builds (with high sccache hits) finishes in under 4 minutes.

Current nightly builds:
Screenshot 2024-08-21 at 10 23 02 AM

Fresh nightly build with the changes:
Screenshot 2024-08-21 at 10 36 12 AM

Nightly build with high `sccache` hit rate:
Screenshot 2024-08-21 at 10 36 33 AM

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,6 +2,16 @@ SHELL:=/bin/bash
 
 PACKAGE=github.com/numaproj/numaflow
 CURRENT_DIR=$(shell pwd)
+
+HOST_ARCH=$(shell arch)
```
Comment (whynowy) [inline Makefile:6]:
```suggestion
HOST_ARCH=$(shell uname -m)
```

Comment (whynowy) [reply Makefile:6]: On Mac intel:

```
arch
i386
```

File: Makefile [hunk 2]
```diff
@@ -75,26 +85,32 @@ dist/$(BINARY_NAME)-linux-ppc64le: GOARGS = GOOS=linux GOARCH=ppc64le
 dist/$(BINARY_NAME)-linux-s390x: GOARGS = GOOS=linux GOARCH=s390x
 
 dist/$(BINARY_NAME):
+	@echo -e "\n------MAKE: $@ ------"
```
Comment (whynowy) [inline Makefile:88]: Can you remove all these logs?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -38,9 +38,58 @@ jobs:
...
+        run: |
+          echo "targets = ['aarch64-unknown-linux-gnu']" >> rust-toolchain.toml
+      - name: Setup Rust toolchain
+        uses: actions-rust-lang/setup-rust-toolchain@v1.9.0
+        with:
+          rustflags: ''
+      - name: Install dependenices
+        run: sudo apt-get install -y gcc-aarch64-linux-gnu protobuf-compiler
+      - name: Build binary
+        run: export CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER='aarch64-linux-gnu-gcc' && RUSTFLAGS='-C target-feature=+crt-static' cargo build --release --target aarch64-unknown-linux-gnu
```
Comment (vigith) [inline .github/workflows/release.yml:81]:
```suggestion
        run: RUSTFLAGS='-C target-feature=+crt-static -C linker=aarch64-linux-gnu-gcc' cargo build --release --target aarch64-unknown-linux-gnu
```

File: .github/workflows/nightly-build.yml [hunk 4]
```diff
@@ -40,9 +40,70 @@ jobs:
...
+      - name: Configure sccache
+        run: |
+          echo "RUSTC_WRAPPER=sccache" >> $GITHUB_ENV
+          echo "SCCACHE_GHA_ENABLED=true" >> $GITHUB_ENV
+      - name: Run sccache-cache
+        uses: mozilla-actions/sccache-action@v0.0.5
+      - name: Install dependenices
+        run: sudo apt-get install -y gcc-aarch64-linux-gnu protobuf-compiler
+      - name: Build binary
+        run: export CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER='aarch64-linux-gnu-gcc' && RUSTFLAGS='-C target-feature=+crt-static' cargo build --release --target aarch64-unknown-linux-gnu
```
Comment (vigith) [inline .github/workflows/nightly-build.yml:95]:
```suggestion
        run: RUSTFLAGS='-C target-feature=+crt-static -C linker=aarch64-linux-gnu-gcc' cargo build --release --target aarch64-unknown-linux-gnu
```

**Review by whynowy — APPROVED** [review]

**Pull request merged**